roachtest: take storage checkpoints on failed consistency checks

erikgrinaker opened the pull request:
**roachprod: add store index parameter for `{store-dir}` expansion**

On multi-store nodes, this allow using e.g. `{store-dir:2}` to reference the second store on the node (NB: not the store ID, but its local index).

**roachprod: expand parameters in `get` source**

This allows using e.g. `{store-dir}` in the source path.

**sql: add `COCKROACH_INTERNAL_CHECK_CONSISTENCY_FATAL`**

This environment variable makes `crdb_internal.check_consistency()` take storage checkpoints and terminate nodes on any inconsistencies when run with `stats_only: false`. This is the same as the background consistency checker, and is done by changing the mode CHECK_FULL to CHECK_VIA_QUEUE.

**roachtest: take storage checkpoints on failed consistency checks**

This allows us to debug range inconsistencies.

**roachtest: collect Pebble checkpoint artifacts**

Resolves #115770.
Epic: none
Release note: None

erikgrinaker committed:
roachtest: take storage checkpoints on failed consistency checks

This allows us to debug range inconsistencies.

Epic: none
Release note: None

erikgrinaker committed:
roachtest: collect Pebble checkpoint artifacts

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/117235)
<!-- Reviewable:end -->

erikgrinaker commented:
@renatolabs @pav-kv Submitting this for an initial look. Will need a few tests and tweaks.

pav-kv commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -1585,6 +1585,59 @@ func (c *clusterImpl) FetchCores(ctx context.Context, l *logger.Logger) error {
+			// Fetch checkpoints.
+			if len(checkpointNodes) > 0 {
+				src := filepath.Join(storePath, "auxiliary", "checkpoints")
+				dest := filepath.Join(c.t.ArtifactsDir(), "checkpoints", fmt.Sprintf("data%d", storeIdx))
```
Can `storeIdx` dup across nodes? Consider putting the node ID into the path too, e.g. like `"n%ds%d"`.

erikgrinaker replied:
>Can storeIdx dup across nodes?

Yes.

`Get()` will handle creating directories for different nodes when fetching from multiple nodes. I.e. if we're fetching ´data1` from node 3,5,7, it will create the directories `3.data1`, `5.data1`, and `7.data1`.

pav-kv commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1399,6 +1399,9 @@ func (r *testRunner) collectArtifacts(
 		if err := c.CopyRoachprodState(ctx); err != nil {
 			t.L().Printf("failed to copy roachprod state: %s", err)
 		}
+		if err := c.FetchPebbleCheckpoints(ctx, t.L()); err != nil {
+			t.L().Printf("failed to fetch Pebble checkpoints: %s", err)
+		}
```
Have you tested that this works?

Could probably use the `inconsistency` roachtest for that.

erikgrinaker replied:
Only manually. I'll add tests once we agree on the direction.

pav-kv reviewed (commented):
The approach looks good

erikgrinaker commented on pkg/kv/kvpb/api.proto:
```diff
@@ -828,6 +828,15 @@ enum ChecksumMode {
+  //
+  // TODO(erikgrinaker): the checkpointing should be moved out of the method,
+  // see TODO on CHECK_VIA_QUEUE and Replica.CheckConsistency.
+  bool checkpoint = 6;
```
It will break in mixed version clusters. I'd like to backport this to 23.2 for additional test coverage, if possible.

The TODOs in this area also point towards moving this kind of logic out of `ChecksumMode`, so I think it makes sense for the caller to specify different combinations of options as needed.

erikgrinaker commented:
@renatolabs Do you have any thoughts on this before I wrap it up?

renatolabs commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -1585,6 +1585,59 @@ func (c *clusterImpl) FetchCores(ctx context.Context, l *logger.Logger) error {
+			// Find any checkpoints.
+			//
+			// TODO(erikgrinaker): expose and use SyncedCluster.NodeDir(). We can't
+			// use the {store-dir} expansion which only refers to the first store.
```
Another option would be for `{store-dir}` to take a store index (more consistent with how similar functionality is exposed).

erikgrinaker replied:
Yeah, may as well -- I'll have a look.

erikgrinaker replied:
Done.

renatolabs reviewed (commented):
Nice!

Left minor comments for consideration. The direction here make sense to me.

erikgrinaker commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -1585,6 +1585,59 @@ func (c *clusterImpl) FetchCores(ctx context.Context, l *logger.Logger) error {
+			storePath := fmt.Sprintf("/mnt/data%d/cockroach", storeIdx)
+			var checkpointNodes option.NodeListOption
+			results, err := c.RunWithDetails(ctx, l, c.All(), "sudo", "/bin/bash", "-c",
+				fmt.Sprintf("'find %s/auxiliary -maxdepth 1 -type d -name checkpoints'", storePath))
```
>I believe this would eliminate the need to check for stdout

Yeah, that's better, thanks. I originally started out with a `find /mnt` here, to try to automagically handle all stores, so the code is a bit weird.

>Also curious: what error do we get without sudo?

Permission denied on `lost+found`, which only `root` has access to. But that was from a previous incarnation, it shouldn't be necessary now that we use `auxiliary` as the base. I'll remove it.

erikgrinaker commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -1585,6 +1585,59 @@ func (c *clusterImpl) FetchCores(ctx context.Context, l *logger.Logger) error {
+// divergence, and only contain data for the inconsistent ranges.
+func (c *clusterImpl) FetchPebbleCheckpoints(ctx context.Context, l *logger.Logger) error {
+	// Don't grab checkpoints on empty clusters or local runs.
+	if c.spec.NodeCount == 0 || c.IsLocal() {
```
Mostly because we don't need it on local runs -- the stores are presumably still around in that case (?).

renatolabs replied:
If running on our own machines, yes, but If the check failed in a TeamCity run, they wouldn't be available.

erikgrinaker replied:
Ah, gotcha.

erikgrinaker commented on pkg/cmd/roachtest/roachtestutil/validation_check.go:
```diff
@@ -46,13 +50,15 @@ func CheckReplicaDivergenceOnDB(ctx context.Context, l *logger.Logger, db *gosql
 SET statement_timeout = '20m';
 SELECT t.range_id, t.start_key_pretty, t.status, t.detail
-FROM crdb_internal.check_consistency(false, '', '') as t;`)
+FROM crdb_internal.check_consistency(false, '', '', true) as t;`)
```
On second thought, this is a bit tricky since we'll likely backport the change, so we'll have to check for a patch version which we don't really know yet.

The consistency check will just log an error if this should happen, without failing the test -- maybe that's fine?

https://github.com/cockroachdb/cockroach/blob/f17c6a849f3a6de8d0204adb4f46025e0f3a6538/pkg/cmd/roachtest/roachtestutil/validation_check.go#L70-L76

renatolabs replied:
Huh interesting.

> Ideally this would be rock solid.

Do you know if things have changed since this comment was added? I wonder what value we get from just logging the error; in all likelihood, legitimate errors would go unnoticed in CI runs.

erikgrinaker commented:
On second thought, I wonder if we should just do something simpler here, which won't require any API changes: add an envvar (e.g. `COCKROACH_INTERNAL_CHECK_CONSISTENCY_VIA_QUEUE`) which will change `crdb_internal.check_consistency()` to use `VIA_QUEUE`, causing it to take checkpoints and terminate nodes if it finds any inconsistencies.

Wdyt @pav-kv?

erikgrinaker committed:
roachprod: add store index parameter for `{store-dir}` expansion

On multi-store nodes, this allow using e.g. `{store-dir:2}` to reference
the second store on the node (NB: not the store ID, but its local
index).

Epic: none
Release note: None

pav-kv commented:
@erikgrinaker 
The current implementation takes checkpoints and doesn’t kill the node, right? If we flip to `VIA_QUEUE`, it’ll kill the node. Is it ok? If not, `COCKROACH_INTERNAL_CHECK_CONSISTENCY_VIA_QUEUE` would need to be honoured both to set `VIA_QUEUE`, and to modify its interpretation. This happens on 2 different nodes (the sender and the receiver below raft).

erikgrinaker commented:
I figured terminating the node seemed fine, since we were going to fail the test anyway.

I guess that prevents us from collecting debug.zip data from the node, so maybe we do want to keep it running -- but if the inconsistency happened to be caught by the normal consistency queue, then we'd be in this situation anyway.

It just seems like a much more light-weight change -- I'm not sure I like modifying the public APIs just for this.

pav-kv commented:
If terminating is fine, then perhaps an env var is fine. Who would set the var?

erikgrinaker commented:
We'd have to set it for all roachtests.

pav-kv commented:
Can we use some of the static constants instead of env var? Like the `buildutil.CrdbTestBuild`. So that all tests/roachtests automatically change the meaning of a full check.

erikgrinaker commented:
Roachtests use arbitrary binaries, in principle don't control whether they're built with `CrdbTestBuild` or otherwise (consider e.g. running `roachtest` locally).

> all tests/roachtests automatically change the meaning of a full check

I don't think we want to do this. For example, it will break all tests of `crdb_internal.check_consistency()`, since they'll unconditionally fatal.

pav-kv commented:
> I don't think we want to do this. For example, it will break all tests of crdb_internal.check_consistency(), since they'll unconditionally fatal.

We may choose to never fatal nodes in tests. Similarly to how to [mock out](https://github.com/cockroachdb/cockroach/blob/e21c0cac5c4d30e28ed619b06c7fd097be86ed77/pkg/kv/kvserver/replica_consistency.go#L867-L872) the fatal behaviour in unit tests.

Also, it doesn't have to be a build flag. Can be any semi-static indicator that we're in a roachtest, similar to testing knobs for the unit tests.

erikgrinaker commented:
>Also, it doesn't have to be a build flag. Can be any semi-static indicator that we're in a [specific] roachtest, similar to testing knobs for the unit tests.

How can we do this? Roachtests use an arbitrary CRDB binary, they can only control runtime flags like envvars, not build-time flags.

pav-kv commented:
Right, then env var is the answer. I was thinking of some "code level" env vars basically, but it appears we don't have them.

erikgrinaker committed:
sql: add `COCKROACH_INTERNAL_CHECK_CONSISTENCY_FATAL`

This environment variable makes `crdb_internal.check_consistency()` take
storage checkpoints and terminate nodes on any inconsistencies when run
with `stats_only: false`. This is the same as the background consistency
checker, and is done by changing the mode CHECK_FULL to CHECK_VIA_QUEUE.

Epic: none
Release note: None

erikgrinaker commented:
Ok, I've added `COCKROACH_INTERNAL_CHECK_CONSISTENCY_FATAL` instead, and adapted the code. I need to add a test for the checkpoint artifacts collection and have a final look over everything, but otherwise I think this should be good now.

erikgrinaker committed:
roachprod: expand parameters in `get` source

This allows using e.g. `{store-dir}` in the source path.

Epic: none
Release note: None

erikgrinaker commented:
This should be ready for final review now. Became a fair bit simpler.

>I need to add a test for the checkpoint artifacts collection

Doesn't look like we have a test harness in place for roachtest artifacts collection, and I don't want to start building one now, so I dropped this. Verified manually with the `inconsistency` test that this works as expected.

pav-kv reviewed (approved):
Nice, thank you!

erikgrinaker commented:
@renatolabs I don't think there should be anything controversial here, so I'll merge this, but let me know if you have further comments and I'll address them separately.

TFTRs!

bors r+
